### ISSUE TITLE
ua issues when running it behind a proxy

### ISSUE THREAD

**Issue by pelacables** [dnanexus/dx-toolkit#581]
Hello,

we're seeing a problem when running ua behind a proxy.
We run version 1.5.31 precompiled.

Test shows everything is ok: (I've removed some private info like project / username and real proxy name / ip. It will be zscaler:1.2.3.4):

```bash
~$ ua  --test
Upload Agent Version: 1.5.31
  git version: v0.280.0
  libboost version: 1.55.0
  libcurl version: 7.45.0
Upload Agent v1.5.31, environment info:
  API server protocol: https
  API server host:     api.dnanexus.com
  API server port:     443
Your copy of Upload Agent is up to date.
System Messages:

  There are currently no system messages.
Current User: XYZ
Current Project: XYZ (project-XYZ)
Proxy Settings:
  http_proxy: http://zscaler:80
  https_proxy: http://zscaler:80
  HTTP_PROXY: http://zscaler:80
  HTTP_PROXY: http://zscaler:80
Operating System:
  Name:    Linux
  Release: 3.10.0-693.17.1.el7.x86_64
  Version: #1 SMP Thu Jan 25 04:11:40 CST 2018
  Machine: x86_64
CA Certificate: /etc/pki/tls/certs/ca-bundle.crt
Resolving Amazon S3:
  Resolved to 52.216.101.53
Testing connection:
  Successfully contacted google.com over http: (200)
  Successfully contacted google.com over https: (200)

What to do next:
  1. Run the Upload Agent with the -v flag to get verbose output:
    ua -v --test
    ua -v <filename>
  2. Set DX_LIBCURL_VERBOSE environment to 1 and repeat the upload attempt to get libcurl debug logs:
    DX_LIBCURL_VERBOSE=1 ./ua <filename>
```

When we try to upload a file we get connection timeout:

```bash
$ ua -v  --project XYZ upload-me-now
Signature of file upload-me-now matches remote file upload-me-now (file-FjQpvGj06JQ4PpB61qVvGYy8), which is 0% complete. Will resume uploading to it.

Failed to upload Chunk [0 - 0] for local file (upload-me-now). APIServer response for last try: ''

File "upload-me-now" could not be uploaded.

1579271833 PID 308532 ... all files closed.
failed
```

After talking to network support they confirm that the application is using the proxy at the begging (when it talks to api.dnanexus.com) but then, when it has to talk to https://dnanexus-platform-upload-prod.s3.amazonaws.com we are not using the proxy anymore so our connection is bein blocked by the firewall:

a connection to api.dnaxesus.com using the proxy looks like:

```bash
> POST /file-FjQpvGj06JQ4PpB61qVvGYy8/describe HTTP/1.1
Host: api.dnanexus.com
User-Agent: DNAnexus-Upload-Agent/1.5.31/v0.280.0 platform/linux uid/7c06730a-3084245c dxcpp/v0.280.0 dxhttp/v0.280.0
Accept: */*
Authorization: bearer BON6C902iwof8plVImxe
Content-Type: application/json; charset=utf-8
DNAnexus-API: 1.0.0
Content-Length: 2

* upload completely sent off: 2 out of 2 bytes
< HTTP/1.1 200 OK
< Server: nginx
< Date: Fri, 17 Jan 2020 14:40:55 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 410
< Connection: keep-alive
< X-Powered-By: Express
< X-Request-ID: 1579272055506-241195
< X-Content-Type-Options: nosniff
<
* Connection #0 to host zscaler left intact
*   Trying 1.2.3.4...
* Connected to zscaler (1.2.3.4) port 80 (#0)
* Establish HTTP proxy tunnel to api.dnanexus.com:443
> CONNECT api.dnanexus.com:443 HTTP/1.1
Host: api.dnanexus.com:443
User-Agent: DNAnexus-Upload-Agent/1.5.31/v0.280.0 platform/linux uid/7c06730a-3084245c dxcpp/v0.280.0 dxhttp/v0.280.0
Proxy-Connection: Keep-Alive

< HTTP/1.1 200 Connection Established
< Proxy-Agent: Zscaler/5.7
<
* Proxy replied OK to CONNECT request
[...]
```
The important line here is:

```bash
* Connected to zscaler (1.2.3.4) port 80 (#0)
```

where it resolves zscaler to 1.2.3.4 properly.

But, later on, when it tries to upload the file to dnanexus-platform-upload-prod.s3.amazonaws.com we're not using the proxy anymore:

```bash
* upload completely sent off: 61 out of 61 bytes
< HTTP/1.1 200 OK
< Server: nginx
< Date: Fri, 17 Jan 2020 14:42:20 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 692
< Connection: keep-alive
< X-Powered-By: Express
< X-Request-ID: 1579272140937-175899
< X-Content-Type-Options: nosniff
<
* Connection #0 to host zscaler left intact
* Added dnanexus-platform-upload-prod.s3.amazonaws.com:443:52.217.39.44 to DNS cache
* Added dnanexus-platform-upload-prod.s3.amazonaws.com:80:52.217.39.44 to DNS cache
* Hostname was found in DNS cache
*   Trying 52.217.39.44...
* Connected to zscaler (52.217.39.44) port 80 (#0)
* Establish HTTP proxy tunnel to dnanexus-platform-upload-prod.s3.amazonaws.com:443
> CONNECT dnanexus-platform-upload-prod.s3.amazonaws.com:443 HTTP/1.1
Host: dnanexus-platform-upload-prod.s3.amazonaws.com:443
User-Agent: DNAnexus-Upload-Agent/1.5.31/v0.280.0 platform/linux uid/7c06730a-3084245c dxcpp/v0.280.0 dxhttp/v0.280.0
Proxy-Connection: Keep-Alive

* Proxy CONNECT aborted due to timeout
```

The relevant line in this second log is 

```bash
* Connected to zscaler (52.217.39.44) port 80 (#0)
```

_seems_ that curl is not resolving zscaler to 1.2.3.4 and using the s3.amazonaws.com IP as the proxy IP.

a more verbose output:

```bash
* upload completely sent off: 61 out of 61 bytes
< HTTP/1.1 200 OK
< Server: nginx
< Date: Fri, 17 Jan 2020 14:45:57 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 692
< Connection: keep-alive
< X-Powered-By: Express
< X-Request-ID: 1579272357810-534198
< X-Content-Type-Options: nosniff
<
* Connection #0 to host zscaler left intact
[1579272357890 7f8d737fe700] DEBUG: Request completed, responseCode = '200'
[1579272357890 7f8d737fe700] DEBUG: Exiting DXHTTPRequest() successfully
[1579272357890 7f8d737fe700] INFO: Chunk [upload-me-now:0-0 -> file-FjQpvGj06JQ4PpB61qVvGYy8[0], tries=3, data.size=0, compress=false]:
[...]
 Host name extracted from URL ('https://dnanexus-platform-upload-prod.s3.amazonaws.com/filev2/a828/file/open/file-FjQpvGj06JQ4PpB61qVvGYy8/1?X-Amz-Algorithm=AWS4-HMAC-SH[...]''): 'dnanexus-platform-upload-prod.s3.amazonaws.com'
[1579272357890 7f8d737fe700] INFO: getRandomIP()--> forceRefreshDNS = false
[1579272357913 7f8d737fe700] INFO: Found address(es) for host (real name): 's3-1-w.amazonaws.com'
[1579272357913 7f8d737fe700] INFO:      1. Pushing address '52.217.42.188' to ipList
[1579272357913 7f8d737fe700] WARNING: Chunk [upload-me-now:0-0 -> file-FjQpvGj06JQ4PpB61qVvGYy8[0], tries=3, data.size=0, compress=false]: Call to getRandomIP() returned: '52.217.42.188'
[1579272357913 7f8d737fe700] INFO: Chunk [upload-me-now:0-0 -> file-FjQpvGj06JQ4PpB61qVvGYy8[0], tries=3, data.size=0, compress=false]: Upload URL: https://dnanexus-platform-upload-prod.s3.amazonaws.com/filev2/a828/
[1579272357913 7f8d737fe700] INFO: Chunk [upload-me-now:0-0 -> file-FjQpvGj06JQ4PpB61qVvGYy8[0], tries=3, data.size=0, compress=false]: Adding ip '52.217.42.188' to resolve list for hostname 'dnanexus-platform-upload-prod.s3.amazonaws.com'
[1579272357913 7f8d737fe700] INFO: Chunk [upload-me-now:0-0 -> file-FjQpvGj06JQ4PpB61qVvGYy8[0], tries=3, data.size=0, compress=false]: Starting curl_easy_perform...
* Added dnanexus-platform-upload-prod.s3.amazonaws.com:443:52.217.42.188 to DNS cache
* Added dnanexus-platform-upload-prod.s3.amazonaws.com:80:52.217.42.188 to DNS cache
* Hostname was found in DNS cache
*   Trying 52.217.42.188...
* Connected to zscaler (52.217.42.188) port 80 (#0)
* Establish HTTP proxy tunnel to dnanexus-platform-upload-prod.s3.amazonaws.com:443
> CONNECT dnanexus-platform-upload-prod.s3.amazonaws.com:443 HTTP/1.1
Host: dnanexus-platform-upload-prod.s3.amazonaws.com:443
User-Agent: DNAnexus-Upload-Agent/1.5.31/v0.280.0 platform/linux uid/3eaa7c22-3f2c706e dxcpp/v0.280.0 dxhttp/v0.280.0
Proxy-Connection: Keep-Alive
```

We are a bit lost right now, but we'd like to ask you if this last "connected to" showing a wrong Proxy -> resolved IP is the source of our issue. Is curl having problems with it DNS cache?

[*] I'm trying to rebuild ua in order to disable curl DNS caching and/or adding an extra curl --resolve option for our proxy  but I'm haiving problem when building curl. What is the exact OpenSSL version needed? (I already tried adding curl resolve options to ~/.curlrc but they are not being pick up.)

Please, let me know if you need further details.

**Comment by pdexheimer** [comment]
Are there any updates on this issue? It's been pinned, so is clearly considered to be important on some level, but no activity is visible since January.

Using 1.5.32, I'm seeing very similar behavior - ua respects the https_proxy environment variables when creating the files in the DNANexus project, but seems to ignore the proxy when attempting to upload the files to S3.

This represents a very big barrier to me actually using the platform, as it means I have to transfer my files to an intermediate location before I can upload them to their final home in DNANexus.

**Comment by pelacables** [comment]
no updates and no patch. They could not reproduce the error.
In our case we ended modifying the source code and hardcoding our proxy in the internal DNS. 
If you are interested in the patch let me know.

**Comment by pdexheimer** [comment]
Just to close the loop on this issue - the PR noted above (DEVEX-1796) has been merged and released as ua v1.5.33.  At least for me, this version resolves this issue and I can now upload successfully from behind a proxy
